### PR TITLE
Fixing caching to permit running e2e and unit tests at same time

### LINKED ISSUE #482: All tests should pass when all tests are ran
## Summary
Two tests fail when ran concurrently with the e2e test:
- `packages/lambda-handler/tests/test_lambda_handler.py`
```python
class TestCreateS3Client:
    def test_create_s3_client(self, moto_setup):
        """Test create S3 client."""
        s3_client = lambda_handler.create_s3_client()
        assert s3_client.meta.endpoint_url == "https://s3.amazonaws.com"
        assert s3_client.meta.region_name == "us-east-1"
        assert s3_client._get_credentials().secret_key == "test_secret_access_key"  # noqa: S105
        assert s3_client._get_credentials().access_key == "test_access_key_id"
```
- `packages/lambda-handler/tests/test_lambda_handler.py`
```
class TestCreateAWSAuth:
    def test_create_aws_auth(self, moto_setup):
        """Test create AWS auth."""
        auth = lambda_handler.create_aws_auth()

        assert auth.access_id == "test_access_key_id"
        assert auth.region == "us-east-1"
```
### Pytest Output
```
===================================================== FAILURES ======================================================
_____________________________________ TestCreateS3Client.test_create_s3_client ______________________________________

self = <test_lambda_handler.TestCreateS3Client object at 0x1528849e0>
moto_setup = <botocore.client.S3 object at 0x15a94c680>

    def test_create_s3_client(self, moto_setup):
        """Test create S3 client."""
        s3_client = lambda_handler.create_s3_client()
        assert s3_client.meta.endpoint_url == "https://s3.amazonaws.com"
        assert s3_client.meta.region_name == "us-east-1"
>       assert s3_client._get_credentials().secret_key == "test_secret_access_key"  # noqa: S105
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       AssertionError: assert 'FOOBARSECRET' == 'test_secret_access_key'
E         
E         - test_secret_access_key
E         + FOOBARSECRET

packages/lambda-handler/tests/test_lambda_handler.py:14: AssertionError
______________________________________ TestCreateAWSAuth.test_create_aws_auth _______________________________________

self = <test_lambda_handler.TestCreateAWSAuth object at 0x152887260>
moto_setup = <botocore.client.S3 object at 0x15b4458b0>

    def test_create_aws_auth(self, moto_setup):
        """Test create AWS auth."""
        auth = lambda_handler.create_aws_auth()
    
>       assert auth.access_id == "test_access_key_id"
E       AssertionError: assert 'FOOBARKEY' == 'test_access_key_id'
E         
E         - test_access_key_id
E         + FOOBARKEY

packages/lambda-handler/tests/test_lambda_handler.py:105: AssertionError
================================================= warnings summary ==================================================
e2e/test_e2e.py::TestEndToEndSimulated::test_upload_and_process
  /Users/jnygaard/Dev/Skylight/Dibbs/dibbs-text-to-code/.venv/lib/python3.12/site-packages/requests_aws4auth/aws4signingkey.py:85: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    self.date = date or datetime.utcnow().strftime('%Y%m%d')

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================== short test summary info ==============================================
FAILED packages/lambda-handler/tests/test_lambda_handler.py::TestCreateS3Client::test_create_s3_client - AssertionError: assert 'FOOBARSECRET' == 'test_secret_access_key'
FAILED packages/lambda-handler/tests/test_lambda_handler.py::TestCreateAWSAuth::test_create_aws_auth - AssertionError: assert 'FOOBARKEY' == 'test_access_key_id'
===================================== 2 failed, 310 passed, 1 warning in 7.04s ======================================

```
This appears to be because `boto3` there is not being isolated or properly cleaned-up across tests.

### PR DESCRIPTION
**Pull request by robertandremitchell**
## Description

Fixes a bug where running `uv run pytest` or `uv run pytest e2e/ packages/lambda-handler` broke because of a caching issue. lambda-handler now checks whether 
## Related Issues

Closes #482 

## Additional Notes

[Add any additional context or notes that reviewers should know about.]

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist

Please review and complete the following checklist before submitting your pull request:

- [ ] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [ ] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [ ] I have updated the documentation, if applicable.
- [ ] I have added or updated test cases to cover my changes, if applicable.
- [ ] I have minimized the number of reviewers to include only those essential for the review.

## Checklist for Reviewers

Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: packages/lambda-handler/src/lambda_handler/lambda_handler.py [hunk 1]
```diff
@@ -15,7 +15,9 @@
 from .models import OpenSearchResult
 
 _cached_aws_auth: AWS4Auth | None = None
+_cached_aws_auth_key: tuple[str | None, str | None, str | None] | None = None
 _cached_s3_client: BaseClient | None = None
+_cached_s3_client_key: tuple[str | None, str | None, str | None, str | None] | None = None
```
Comment (JNygaard-Skylight) [inline packages/lambda-handler/src/lambda_handler/lambda_handler.py]: I think this is more of a nit, but this typing is pretty hard to read. Does this have to be a tuple of three optional strings?  Seems like it would be cleaner for it to be a class, or even just a dictionary, but I'm not as familiar with the AWS client stuff. Perhaps this would be a good place to use a type alias?

**Review by JNygaard-Skylight — APPROVED** [review]
Looks good! Just a nit about the typing, but I think that might be more of a personal problem on my part.

File: packages/lambda-handler/src/lambda_handler/lambda_handler.py (hunk 1 above)
Comment (robertandremitchell) [reply packages/lambda-handler/src/lambda_handler/lambda_handler.py]: I agree. I took a look and i think just adding explicit reset logic might end up being a little cleaner here? Curious your thoughts!

**Review by JNygaard-Skylight — APPROVED** [review]
LGTM!
